src/strings/ops.c

```diff
@@ -468,16 +471,36 @@ MVMString * MVM_string_concatenate(MVMThreadContext *tc, MVMString *a, MVMString
     if (bgraphs == 0)
         return a;
 
+    is_concat_stable = MVM_nfg_is_concat_stable(tc, a, b);
+    if (is_concat_stable == 0) {
+        MVMCodepoint last_a_first_b[2];
+        last_a_first_b[0] = MVM_string_get_grapheme_at_nocheck(tc, a, a->body.num_graphs - 1);;
+        last_a_first_b[1] = MVM_string_get_grapheme_at_nocheck(tc, b, 0);
+        /* Needing both to be CCC = 0 can probably be relaxed some, but be careful optimizing */
+        if (0 <= last_a_first_b[0] && 0 <= last_a_first_b[1] && MVM_unicode_relative_ccc(tc, last_a_first_b[0]) == 0 && MVM_unicode_relative_ccc(tc, last_a_first_b[1]) == 0) {
+            renormalized_section = MVM_unicode_codepoints_c_array_to_nfg_string(tc, last_a_first_b, 2);
+            MVM_gc_root_temp_push(tc, (MVMCollectable **)&(renormalized_section));
```

Comment:
(Yes, it's probably needed later, but could be done later as a result, perhaps allowing the use of the easier-to-check `MVMROOT`.)